### ISSUE TITLE
Question about multi-output GPR 

### ISSUE THREAD

**Issue by coderxulh** [EmuKit/emukit#439]
Hi, everyone,
I'm working to find a way for multi-output GP with multi-fidelity model. I wonder if GPyMultiOutputWrapper in emukit can be used to solve it? If then, how can I start from the original Nonlinear multi-fidelity model (NARGP) ?
I just got confused with GPyMultiOutputWrapper used in Linear multi-fidelity modeling notebook. Is that just a multi-output(or multi-task) GP to model the correlation between the two fidelities(tasks)? That seems to be different from the general coregionalisation model?

**Comment by coderxulh** [comment]
Hi, apaleyes~ Thank you for your answers. It seems that your multi-output for multi-fidelity in Emukit is a specifically pre-defined correlation between two fidelities with some hyperparameters, which may not be suitable for a commonly used multi-output modelling.
Also, I wonder if a commonly used multi-output regression method like in GPy [https://nbviewer.org/github/SheffieldML/notebook/blob/master/GPy/coregionalized_regression_tutorial.ipynb](url), GPflow [https://gpflow.github.io/GPflow/2.7.1/notebooks/advanced/multioutput.html](url) and [https://docs.gpytorch.ai/en/latest/examples/03_Multitask_Exact_GPs/Multitask_GP_Regression.html](url) in your schedule?

**Comment by apaleyes** [comment]
Unfortunately we don't have any further extensions of our multi-fidelity package planned at the moment. Although PRs are always welcome!

**Comment by AlbanMor** [comment]
Hi there, 

I have another question that fits under this title:

Is there a way to use a "GPyMultiOutputWrapper" model (that wrapped a GPy.models.MultioutputGP) for an "Experimental Design"? 

It would for example suggest points where the combined uncertainty is maximum? 

The wrapper specifies that "X inputs should have the corresponding output index as the last column in the array". But in the case of the Emukit Experimental Design and BO the X that is passed to the model is defined by the parameter space correct?

If you could point me towards a way to make it work that would be great! 

Thank you!

**Comment by apaleyes** [comment]
In theory there is nothing that prevents you from using GPyMultiOutputWrapper for experimental design. This wrapper exists to support multi-fidelity use cases, so it all depends on the nature of your data. "the X that is passed to the model is defined by the parameter space correct"  - that is correct and defined by your objective. So if you are dealing with let's say two simulations, one is a more precise version of the other, then you should be able to use GPyMultiOutputWrapper  with it.

**Comment by AlbanMor** [comment]
@apaleyes 2, Thank you so much for taking the time to reply. Below is my toy example trying to leverage GPyMultiOutputWrapper for experimental design. 

I will let you look into it and tell me how i'm using it wrong but I feel like the issue is that the wrapped model is expecting [one dimensional input + the index referring to one of the 2 outputs], but the parameter space is only "pushing" the value of the one dimension input. Ultimately we would want the experimental design loop to probe both outputs for the given input and combine somehow the uncertainty of its prediction on both outputs. 

Here is the code, your help would be greatly appreciated:

#begining of code is heavily inspired from GPY tutorial: https://nbviewer.org/github/SheffieldML/notebook/blob/master/GPy/coregionalized_regression_tutorial.ipynb

import GPy
import numpy as np
import matplotlib.pyplot as plt

## Setuping the GPy moodel

# 1 input dimension
# 2 outputs

#This functions generate data corresponding to two outputs
f_output1 = lambda x: 4. * np.cos(x/5.) - .4*x - 35. + np.random.rand(x.size)[:,None] * 2.
f_output2 = lambda x: 6. * np.cos(x/5.) + .2*x + 35. + np.random.rand(x.size)[:,None] * 8.


#{X,Y} training set for each output
X1 = np.random.rand(100)[:,None]; X1=X1*75
X2 = np.random.rand(100)[:,None]; X2=X2*70 + 30
Y1 = f_output1(X1)
Y2 = f_output2(X2)
#{X,Y} test set for each output
Xt1 = np.random.rand(100)[:,None]*100
Xt2 = np.random.rand(100)[:,None]*100
Yt1 = f_output1(Xt1)
Yt2 = f_output2(Xt2)


#let's create seperate models for each output (the point to this is to be able to easily use tailored kernels to predict each outputs)
K = GPy.kern.Matern32(1)

m1 = GPy.models.GPRegression(X1,Y1,kernel=K.copy())
m1.optimize()
m2 = GPy.models.GPRegression(X2,Y2,kernel=K.copy())
m2.optimize()

fig = plt.figure(figsize=(12,8))
xlim = (0,100)#; ylim = (0,50)
#Output 1
ax1 = fig.add_subplot(211)
m1.plot(plot_limits=xlim,ax=ax1)
ax1.plot(Xt1[:,:1],Yt1,'rx',mew=1.5)
ax1.set_title('Output 1')
#Output 2
ax2 = fig.add_subplot(212)
m2.plot(plot_limits=xlim,ax=ax2)
ax2.plot(Xt2[:,:1],Yt2,'rx',mew=1.5)
ax2.set_title('Output 2')

#now we will wrap both models into one multiouput model

m3=GPy.models.MultioutputGP(X_list=[m1.X,m2.X],
                            Y_list=[m1.Y,m2.Y],
                            kernel_list=[m1.kern,m2.kern],
                            likelihood_list=[m1.likelihood,m2.likelihood])



## Experimental design to efficiently explore the parameter space and decrease the overall model uncertainty on its prediction

#1 input dimension
from emukit.core import ParameterSpace, ContinuousParameter
parameter_space = ParameterSpace([ContinuousParameter('x1', 0, 100)])


#Wrapping the multioupt GPy model
from emukit.model_wrappers import GPyMultiOutputWrapper
model_emukit = GPyMultiOutputWrapper(gpy_model=m3,n_outputs=2,n_optimization_restarts=10)

model_emukit.optimize()

#probing the model
from emukit.experimental_design.acquisitions import ModelVariance
model_variance = ModelVariance(model = model_emukit)

from emukit.experimental_design import ExperimentalDesignLoop
expdesign_loop = ExperimentalDesignLoop(model = model_emukit,
                                         space = parameter_space,
                                         acquisition = model_variance,
                                         batch_size = 5)
results=None
next_points=expdesign_loop.get_next_points(results)

**Comment by apaleyes** [comment]
> Ultimately we would want the experimental design loop to probe both outputs for the given input and combine somehow the uncertainty of its prediction on both outputs.

I guess this is where you are going wrong. The way MF works in Emukit is that at every step it chooses which fidelity to probe. And the main idea is to save time/money/effort running high fidelity function where low (and cheaper) fidelity would do. What you are describing is interesting, and can probably be useful when one wants to quantify the gap between the fidelities, but we haven't explored this in Emukit yet. It's just a matter of writing an acquisition function that can do those things though, if you feel like doing it.

Code-wise you are exactly right, parameter space only defines one parameter, input variable itself, but shall also contain a parameter that informs the model about the fidelity level.
